File: src/main/scala/2017/day-16.scala
package adventofcode2017

import scala.io.Source

object Day16 {
  lazy val moves: String = Source.fromFile(s"2017/input/day-16.input").mkString.trim
  val input = "abcdefghijklmnop".toArray
  
  val actions: Map[Char, (Array[Char], String, String) => Array[Char]] = Map[Char, (Array[Char], String, String) => Array[Char]] (
    's' -> { case (input, x, _) => 
      val cutPoint = input.length - x.toInt;
      input.slice(cutPoint, input.length) ++ input.slice(0, cutPoint)
    },
    'x' -> { case (input, a, b) => 
      val temp = input(a.toInt);
      input(a.toInt) = input(b.toInt)
      input(b.toInt) = temp;
      input
    },
    'p' -> { case (num, a, b) => 
      actions('x')(num, num.indexOf(a(0)).toString, num.indexOf(b(0)).toString);
    }
  )

  def performMove(input: Array[Char], move:String): Array[Char] = {
    val action = move(0);
    val params = move.substring(1).split("/").toArray;
    val f = actions(action)
    f(input, params(0), if (params.length == 2) params(1) else "");
  }

  def findLoopIndex(input: Array[Char], movesOrder: Array[String], cache: Set[String] = Set(input.mkString), times: Int = 0): Int = {
    val newInput = movesOrder.foldLeft(input) {
      (input, move) => performMove(input, move)
    }
    if (cache(newInput.mkString)) times + 1 
    else findLoopIndex(newInput, movesOrder, cache + newInput.mkString, times + 1)
  }

  def main(args: Array[String]): Unit = {
    val movesOrder = moves.split(",");

    val oneDance = movesOrder.foldLeft(input) {
      (input, move) => performMove(input, move)
    }
    println(s"part1: ${oneDance.mkString}")

    val loopIndex = findLoopIndex(input, movesOrder)
    val start = Math.floor(1000000000/loopIndex)*loopIndex;
    val res = (start.toInt until 1000000000).foldLeft(input) {
      (input, _) => movesOrder.foldLeft(input) {
        (input, move) => performMove(input, move)
      }
    }
    println(s"part2: ${res.mkString}")
  }
}

File: 2017/js/day-16.js
const fs = require('fs');
const moves = fs.readFileSync('2017/input/day-16.input').toString();

const actions = {
  s: (input, a) => {
    const rest = input.splice(input.length - a);
    input = [...rest, ...input];
    return input;
  },
  x: (input, a, b) => {
    const temp = input[a];
    input[a] = input[b]
    input[b] = temp;
    return input;
  },
  p: (input, a, b) => {
    return actions.x(input, input.indexOf(a), input.indexOf(b));
  }
};

const getInput = () => 'abcdefghijklmnop'.split('');

const performMove = (input, move) => {
  const action = move[0];
  const params = move.substring(1).split('/');
  return actions[action](input, params[0], params[1]);
}
const movesOrder = moves.split(',');
let oneDance = getInput();
movesOrder.forEach(move => 
  oneDance = performMove(oneDance, move));
console.log(`part1: ${oneDance.join('')}`);

let searchInput = getInput();
let cache = {}; // finding the loop
let loopIndex = -1;
for(let i=0; i<1000000000; i++) {
  movesOrder.forEach(move => 
    searchInput = performMove(searchInput, move));
  if (cache[searchInput]) {
    loopIndex = i;
    break;
  } else {
    cache[searchInput] = true;
  }
}

let input = getInput();
const start = Math.floor(1000000000/loopIndex)*loopIndex;
for(let i=start; i<1000000000; i++) {
  movesOrder.forEach(move => 
    input = performMove(input, move));
}
console.log(`part2: ${input.join('')}`);


